flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -281,10 +289,41 @@ protected void processInput(DefaultActionContext context) throws Exception {
 		if (status == InputStatus.END_OF_INPUT) {
 			context.allActionsCompleted();
 		}
-		else if (status == InputStatus.NOTHING_AVAILABLE) {
+
+		CompletableFuture<?> jointFuture = getInputOutputJointFuture(status);
+		if (jointFuture != null) {
 			SuspendedMailboxDefaultAction suspendedDefaultAction = context.suspendDefaultAction();
-			inputProcessor.getAvailableFuture().thenRun(suspendedDefaultAction::resume);
+			jointFuture.thenRun(suspendedDefaultAction::resume);
+		}
+	}
+
+	/**
+	 * @return a combination of input and output futures if at-least one future of them is not
+	 * completed, otherwise return null if all input and outputs are available.
+	 */
+	private CompletableFuture<?> getInputOutputJointFuture(InputStatus status) {
+		if (status == InputStatus.MORE_AVAILABLE && isOutputAvailable()) {
+			return null;
+		}
+
+		int length = recordWriters.size();
+		for (int i = 0; i < length; i++) {
+			inputOutputFutures[i] = recordWriters.get(i).getAvailableFuture();
+		}
+		inputOutputFutures[length] = inputProcessor.getAvailableFuture();
+		return CompletableFuture.allOf(inputOutputFutures);
+	}
+
+	/**
+	 * @return true if all the record writers are available.
+	 */
+	private boolean isOutputAvailable() {
+		for (RecordWriter recordWriter : recordWriters) {
```

Comment:
I think we should avoid this array looping. I would expect this to be visible in the benchmarks (similarly how array looping of `int[] selectChannels()` was (marginally) visible.

Maybe create an interface to handle this, with two specialisations: single `RecordWriter` and multiple `RecordWriter`s, that would also encapsulate `private final List<RecordWriter<SerializationDelegate<StreamRecord<OUT>>>> recordWriters;` field? Hiding this type would be worthwhile on it's own.

Optimising single `RecordWriter` case is I think optional, as long as you checked that there is no performance regression (`mapRebalanceMapSink` and `mapSink` benchmarks should be good enough for that). However encapsulating `recordWriters` and `inputOutputFutures` fields together behind some utility class make sense on it's own.

Reply:
Yeah, it makes sense to do that, then we can compare the benchmark performance between single `RecordWriter` and multiple `RecordWriter`s. Also it would thin the `StreamTask` a bit to hide two arrays.

Reply:
After reviewing the current usages of `List<RecordWriter<SerializationDelegate<StreamRecord<OUT>>>> recordWriters`, it seems not easy to hide it inside the new proposed interface, mainly because it is used in `OperatorChain` for looping to create the respective `RecordWriterOutput`. And if we adjust the `RecordWriterOutput` accordingly, it would further involve in series of more changes. 

Therefore I suggest retaining the previous array of `RecordWriter`s in `StreamTask` and only proposing a new interface `OutputAvailability` for implementing the specific `SingleOutputAvailability` and `MultipleOutputAvailability`. Then we could avoid always having the array of futures in `StreamTask` which might cause extra overhead for single case. WDYT?

Reply:
Can not we define the following class:
```
/** wrapper class encapsulating multiple RecordWriters */
class RecordWriters<OUT> {
  RecordWriter<SerializationDelegate<StreamRecord<OUT>>>[] recordWriters;
  CompletableFuture<?>[] outputAvailability;

  public RecordWriter<...> getRecordWriter(int outputIndex) { return recordWriters[outpuIndex];}
 
  public CompletableFuture<?> isAvailable() { ... } ;

  public void close() { // loop and close... } 

  public void broadcastEvent(AbstractEvent event) { // loop and broadcast } ;
}
```
?
It seems like we could use it directly in all of the places, both in `StreamTask` and pass it to the `OperatorChain`.

Reply:
Yeah, it makes sense. I assumed it too complicated by mistake before.